### PR TITLE
compiler/wasm: memoize in function (not callsite)

### PR DESCRIPTION
**Pull request by srenatus**
As soon as a function is called twice, there should be a little bit
to gain from having the memoization happen in the function body.

Run on a big bundle, this results in the following change in opcode
counts:

        Total opcodes: 184881 -> 183852

        Opcode counts:
        local.get: 46920 -> 46803 -
        i32.const: 24152 -> 23708 -
        local.set: 14097 -> 14097 =
        call:      12547 -> 12103 -
        end:       10911 -> 11016 +
        br_if:     11062 -> 10840 -
        block:      8987 ->  9092 +

So, the change isn't dramatic at all. But less is more, so I guess
we could still do this.

### CODE REVIEW

**Comment by srenatus** [comment]
😬 forgot about something: the args length check.

**Comment by srenatus** [comment]
Also forgot that we can drop the memo around call_indirect, and thereby the elem_to_func mechanism. ✨

**Comment by srenatus** [comment]
✔️  The `call_indirect` path should be taken care of now. As an implication, we'll drop the extra data section for mapping elem idx -> func idx, together with the lookup function. So, the overall module gets a little lighter.

File: internal/compiler/wasm/wasm.go [hunk 1]
```diff
@@ -635,15 +628,37 @@ func (c *Compiler) compileFunc(fn *ir.Func) error {
...
+		} else { // last block, no wrapping
+			// memoization: insert, spliced into the instructions right
+			// before the return:
+			for _, instr := range instrs {
+				if _, ok := instr.(instruction.Return); ok && memoize {
+					c.appendInstr(instruction.I32Const{Value: int32(idx)})
+					c.appendInstr(instruction.GetLocal{Index: c.local(fn.Return)})
+					c.appendInstr(instruction.Call{Index: c.function(opaMemoizeInsert)})
+				}
+				c.appendInstr(instr)
```
Comment (srenatus) [inline internal/compiler/wasm/wasm.go:660]: 💭 Currently, if this instruction **is** `Return`, we could just ignore it. The last value on the stack is the function's return value, with our without that statement.

**Review by tsandall — APPROVED** [review]
LGTM. It's nice how this cuts down the amount of code in the backend because we don't have to special case CallStmt vs CallDynamicStmt.

**Pull request merged**